### PR TITLE
AVLTree Implementation

### LINKED ISSUE #112: Implement AVL Tree for near-sdk-as
An AVL tree is a balanced binary tree [<sup>1</sup>][1], which swaps nodes in the tree as elements are inserted and deleted in order to maintain the minimum height of the tree.  This ensures that the worst case for accessing an element is `O(log n)`, versus `O(n)` for a normal binary tree.

Currently smart contracts on NEAR provide a Key-Value storage but does not provide an API for accessing/iterating over the keys.  Thus this data structure can be used to create an ordered map by inserting into the tree keys as values and their key as an increasing counter of all entries.

[PersistentOrderedMap](https://github.com/near/near-sdk-as/pull/85/files#diff-ed2627c85f93300ebbbb194eb7fb755a) provides an example of iterating over the last `n` elements, but to maintain order it only allows the last entry to be deleted.

Please use an equivalent implementation of AVL Map in Rust for guidance: https://github.com/near/near-sdk-rs/pull/154 It is okay to literally rewrite the above Rust code in AssemblyScript.

## API

```ts

class AVLTree<K,V> {

  /**
   * A string name is used as a prefix for writing keys to storage.
   */
  constructor(name: string){}

  /**
   * Number of elements in the tree.
   */
  get size(): u32;

  /**
   * Returns whether the key is present in the tree.
   */
  has(key: K): boolean;
  //alias to match rust sdk
  containsKey(key: K): boolean;

  /**
   * If key is not present in the tree, a new node is added with the value.
   * Otherwise update the node with the new value.
   */
  set(key: K, value: V): void;
  //alias to match rust sdk
  insert(key: K, value: V): void;

  /**
   * Get value with key and throw if the key is not in the tree.
   */
  get(key: K): V;

  /**
   * Get value with key and return default value if key is not in the tree.
   */
  getSome(key: K, default: V): V;

  /**
   * Delete element with key if present, otherwise do nothing.
   */
  delete(key: K): void;
  //alias to match rust sdk
  remove(key: K): void;


  /**
   * Get a range of values from a start key to an end key exclusive.
   * If end is greater than max key, include start to max inclusive.
   */
  range(start: K, end: K): V[];

  /**
   * Returns minimum key.
   * Throws if tree is empty.
   */
  min(): K;

  /**
   * Returns maximum key.
   * Throws if tree is empty.
   */
  max(): K;

  /**
   * Returns the minimum key that is strictly greater than the key.
   * Throws if empty or if key is lower than `this.min()`.
   */
  lower(key: K): K;

  /**
   * Returns the maximum key that is strictly less than the key.
   * Throws if empty or if key is higher than `this.max()`.
   */
  higher(key: K): K;

  /**
   * Returns the minimum key that is greater or equal than the key.
   * Throws if empty or if key is lower than `this.min()`.
   */
  lowerOrEqual(key: K): K;

  /**
   * Returns the maximum key that is less or equal than the key.
   * Throws if empty or if key is higher than `this.max()`.
   */
  higherOrEqual(key: K): K;
}
```

### Implementation

Other than the API above and the requirement for the same algorithmic characteristics of a AVL tree, the implementation is open to interpretation.

### Testing

There must be tests that cover each edge case. Use [PersistentSet](https://github.com/near/near-sdk-as/blob/master/assembly/__tests__/runtime/persistent-set.spec.ts) as a guide for how to write tests.

Please make sure the implementation contains all tests of AVL tree that were implemented in https://github.com/near/near-sdk-rs/pull/154

[1]: https://en.wikipedia.org/wiki/AVL_tree

### PR DESCRIPTION
**Pull request by kaiba42**
Resolves #112

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -2,5 +2,5 @@ module.exports = {
   preset: 'ts-jest',
   testEnvironment: 'node',
   testMatch: ["**/__tests__/**/*.spec.ts"],
-  testPathIgnorePatterns: ["/assembly/", "/node_modules/"],
```
Comment (willemneal) [inline jest.config.js:5]: Don't do this.  We need jest to ignore this folder.

Comment (kaiba42) [reply jest.config.js]: got it!

File: assembly/__tests__/runtime/avl-tree.spec.ts [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+    // From near-sdk-rs TreeMap:
+    // h <= C * log2(n + D) + B
+    // where:
+    // C =~ 1.440, D =~ 1.065, B =~ 0.328
+    // (source: https://en.wikipedia.org/wiki/AVL_tree)
+    const B: f64 = -0.328;
+    const C: f64 = 1.440;
+    const D: f64 = 1.065;
+
+    const h = C * Math.log2(n + D) + B;
```
Comment (willemneal) [inline assembly/__tests__/runtime/avl-tree.spec.ts:35]: This uses floating point numbers which aren't allowed, so you should use `math.binaryLog` instead, which only uses integers.

Comment (willemneal) [reply assembly/__tests__/runtime/avl-tree.spec.ts:35]: Ah never mind this is valid in a test file, but is true the actual file.

File: assembly/runtime/collections/avlTree.ts [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+
+    /**
+     * Get a range of entries from a start key (inclusive) to an end key (exclusive).
+     * If end is greater than max key, include start to max inclusive.
+     * 
+     * @param start Key for lower bound (inclusive).
+     * @param end Key for upper bound (exclusive).
+     * @returns Range of entries corresponding to keys within start and end bounds.
+     */
+    entries(start: K, end: K): MapEntry<K, V>[] {
```
Comment (willemneal) [inline assembly/runtime/collections/avlTree.ts:227]:
```suggestion
    entries(start: K = this.min(), end: K this.max()): MapEntry<K, V>[] {
```

After writing this, I think we should make the end of the range be inclusive so that by default this will include the max key.  Same goes for values and keys.

Comment (kaiba42) [reply assembly/runtime/collections/avlTree.ts:227]: Ah yes I had the same initial approach, but unfortunately it results in an unexpected behavior where calling entries() will throw instead of returning an empty array when there are no keys in the tree. This happens because this.min() (and this.max()) throws if the tree is empty.

File: assembly/__tests__/avlTreeContract.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import { AVLTree } from "../runtime";
+
```
Comment (willemneal) [inline assembly/__tests__/avlTreeContract.ts:2]: Any top level code is executed at the start before the contract methods are called.  So you can just define it at the top level once.  However, I've found that it's best to import it from another file so that if you wanted to unit test this, the test could also import the tree.  We currently have an issue with unit testing, where even though we wipe away the storage and state of the Mock VM that handles the `env` functions, we don't wipe away the state of the contract.  This means you need access to the same tree object as the "contract" file so you can clear it.  It's fine in your case because since the trees are not in the global scope they won't be used again.

Comment (kaiba42) [reply assembly/__tests__/avlTreeContract.ts:2]: I think I understand. Am I correct in thinking that it's imperative that unit tests and contracts use distinct prefixes when creating a collection (e.g. AVLTree)?

Comment (willemneal) [reply assembly/__tests__/avlTreeContract.ts:2]: This would only matter if you were creating the tree is another file and didn't have access to it, e.g. testing contract methods that interact with the tree.  In this case however, you can just do a `beforeEach` and reconstruct the tree.  The issue is that the persistent collections usually have a length or size field, which when not set reads from the storage, and when the storage is refreshed but the local collection isn't it'll have the incorrect size.  So reconstructing the collection before each test ensures that the length is uninitialized and will agree with the storage.

Comment (willemneal) [reply assembly/__tests__/avlTreeContract.ts:2]: Furthermore, since this contract isn't tested with unit tests, it's fine to just use a global variable for the tree.

**Comment by willemneal** [comment]
I made some changes to the API so that `height` is the root height.

File: assembly/__tests__/avlTreeContract.ts (hunk 4 above)
Comment (kaiba42) [reply assembly/__tests__/avlTreeContract.ts:2]: Ah I see! Changes you made make sense.

**Comment by kaiba42** [comment]
Thanks for all your help! Changes for having inclusive end bounds are pushed. Let me know if there's anything else and I will get to it asap.
